velox/row/UnsafeRowDeserializers.cpp

```diff
@@ -0,0 +1,279 @@
+    const uint8_t* memoryAddress,
+    memory::MemoryPool* pool) {
+  auto column = BaseVector::create<FlatVector<int128_t>>(type, numRows, pool);
+  auto rawValues = column->mutableRawValues<uint8_t>();
+  for (auto row = 0; row < numRows; row++) {
+    if (!isNullAt(memoryAddress + offsets[row], columnIdx)) {
+      const int64_t offsetAndSize =
+          *(int64_t*)(memoryAddress + offsets[row] + fieldOffset);
+      const int32_t length = static_cast<int32_t>(offsetAndSize);
+      const int32_t wordOffset = static_cast<int32_t>(offsetAndSize >> 32);
+      rawValues[row] =
+          UnsafeRowPrimitiveBatchDeserializer::deserializeLongDecimal(
+              std::string_view(reinterpret_cast<const char*>(
+                  memoryAddress + offsets[row] + wordOffset, length)));
+    } else {
+      column->setNull(row, true);
+    }
+  }
+  return column;
+}
+
+template <>
+VectorPtr createFlatVectorFast<TypeKind::BOOLEAN>(
+    const TypePtr& type,
+    column_index_t columnIdx,
+    vector_size_t numRows,
+    int64_t fieldOffset,
+    const std::vector<int64_t>& offsets,
+    const uint8_t* memoryAddress,
+    memory::MemoryPool* pool) {
+  auto column = BaseVector::create<FlatVector<bool>>(type, numRows, pool);
+  auto rawValues = column->mutableRawValues<uint64_t>();
+  for (auto row = 0; row < numRows; row++) {
+    if (!isNullAt(memoryAddress + offsets[row], columnIdx)) {
+      const bool value = *(bool*)(memoryAddress + offsets[row] + fieldOffset);
+      bits::setBit(rawValues, row, value);
+    } else {
+      column->setNull(row, true);
+    }
+  }
+  return column;
+}
+
+template <>
+VectorPtr createFlatVectorFast<TypeKind::TIMESTAMP>(
+    const TypePtr& type,
+    column_index_t columnIdx,
+    vector_size_t numRows,
+    int64_t fieldOffset,
+    const std::vector<int64_t>& offsets,
+    const uint8_t* memoryAddress,
+    memory::MemoryPool* pool) {
+  auto column = BaseVector::create<FlatVector<Timestamp>>(type, numRows, pool);
+  for (auto row = 0; row < numRows; row++) {
+    if (!isNullAt(memoryAddress + offsets[row], columnIdx)) {
+      const int64_t value =
+          *(int64_t*)(memoryAddress + offsets[row] + fieldOffset);
+      column->set(row, Timestamp::fromMicros(value));
+    } else {
+      column->setNull(row, true);
+    }
+  }
+  return column;
+}
+
+template <>
+VectorPtr createFlatVectorFast<TypeKind::VARCHAR>(
+    const TypePtr& type,
+    column_index_t columnIdx,
+    vector_size_t numRows,
+    int64_t fieldOffset,
+    const std::vector<int64_t>& offsets,
+    const uint8_t* memoryAddress,
+    memory::MemoryPool* pool) {
+  auto column = BaseVector::create<FlatVector<StringView>>(type, numRows, pool);
+  const auto totalSize = getTotalStringSize(
+      columnIdx, numRows, fieldOffset, offsets, memoryAddress);
+  char* rawBuffer = column->getRawStringBufferWithSpace(totalSize, true);
+  for (auto row = 0; row < numRows; row++) {
+    if (!isNullAt(memoryAddress + offsets[row], columnIdx)) {
```

Comment:
nit: 'memoryAddress + offsets[row]' is used several times, any chance to make it a variable?

Reply:
I think not. It uses 3 variables and used at hot spot branch, make it to a function does not make code clear.